File: src/no/tornado/tornadofx/idea/TornadoFXConfigurable.java
package no.tornado.tornadofx.idea;

import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.options.SearchableConfigurable;
import com.intellij.ui.HyperlinkAdapter;
import com.intellij.ui.HyperlinkLabel;
import no.tornado.tornadofx.idea.TornadoFXSettings;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.event.HyperlinkEvent;
import java.awt.*;
import java.io.IOException;
import java.net.URI;

public class TornadoFXConfigurable implements SearchableConfigurable {
    private TornadoFXSettings settings;
    private ConfigUI ui;

    public TornadoFXConfigurable(TornadoFXSettings settings) {
        this.settings = settings;
    }

    @NotNull
    @Override
    public String getId() {
        return getHelpTopic();
    }

    @Nullable
    @Override
    public Runnable enableSearch(String option) {
        return null;
    }

    @Nls
    @Override
    public String getDisplayName() {
        return "TornadoFX";
    }

    @Nullable
    @Override
    public String getHelpTopic() {
        return "preferences.TornadoFX";
    }

    @Override
    public boolean isModified() {
        return ui.isModified(settings);
    }

    @Override
    public void disposeUIResources() {
        ui = null;
    }

    @Override
    public void apply() throws ConfigurationException {
        ui.apply(settings);
    }

    @Nullable
    @Override
    public JComponent createComponent() {
        ui = new ConfigUI();
        return ui.mainPanel;
    }

    @Override
    public void reset() {
        ui.reset(settings);
    }

    public static class ConfigUI {
        private JPanel mainPanel;
        private JCheckBox alternativePropertySyntaxCheckbox;
        private HyperlinkLabel propertySyntaxLink;

        public ConfigUI() {
            propertySyntaxLink.setHyperlinkText("More");
            propertySyntaxLink.addHyperlinkListener(new HyperlinkAdapter() {
                @Override
                protected void hyperlinkActivated(HyperlinkEvent e) {
                    try {
                        Desktop.getDesktop().browse(URI.create("https://edvin.gitbooks.io/tornadofx-guide/content/Appendix%20A%20-%20Supplementary%20Topics.html#alternative-property-syntax"));
                    } catch (IOException e1) {
                        e1.printStackTrace();
                    }
                }
            });
        }

        void reset(TornadoFXSettings settings) {
            alternativePropertySyntaxCheckbox.setSelected(settings.getAlternativePropertySyntax());
        }

        void apply(TornadoFXSettings settings) {
            settings.setAlternativePropertySyntax(alternativePropertySyntaxCheckbox.isSelected());
        }

        boolean isModified(TornadoFXSettings settings) {
            return settings.getAlternativePropertySyntax() != alternativePropertySyntaxCheckbox.isSelected();
        }

    }
}

File: src/main/kotlin/no/tornado/tornadofx/idea/TornadoFXConfigurable.kt
package no.tornado.tornadofx.idea

import com.intellij.openapi.options.SearchableConfigurable
import com.intellij.ui.HyperlinkAdapter
import com.intellij.ui.HyperlinkLabel
import java.awt.Desktop
import java.io.IOException
import java.net.URI
import javax.swing.JCheckBox
import javax.swing.JComponent
import javax.swing.JPanel
import javax.swing.event.HyperlinkEvent

class TornadoFXConfigurable : SearchableConfigurable {
    private val id = "preferences.TornadoFX"
    private val displayName = "TornadoFX"
    private val settings by lazy { TornadoFXSettings.getInstance() }
    private var ui: ConfigUI? = null

    override fun createComponent(): JComponent? = ui?.mainPanel ?: ConfigUI().also { ui = it }.mainPanel

    override fun isModified(): Boolean = ui?.isModified(settings) ?: false

    override fun apply() {
       ui?.apply(settings)
    }

    override fun getDisplayName(): String = displayName

    override fun getId(): String = id

    override fun getHelpTopic(): String? = id

    override fun disposeUIResources() {
        ui = null
    }

    override fun reset() {
        ui?.reset(settings)
    }

    class ConfigUI {
        internal var mainPanel: JPanel? = null
        private var alternativePropertySyntaxCheckbox: JCheckBox? = null
        private var propertySyntaxLink: HyperlinkLabel? = null

        fun reset(settings: TornadoFXSettings) {
            alternativePropertySyntaxCheckbox!!.isSelected = settings.alternativePropertySyntax
        }

        fun apply(settings: TornadoFXSettings) {
            settings.alternativePropertySyntax = alternativePropertySyntaxCheckbox!!.isSelected
        }

        fun isModified(settings: TornadoFXSettings): Boolean {
            return settings.alternativePropertySyntax != alternativePropertySyntaxCheckbox!!.isSelected
        }

        init {
            propertySyntaxLink!!.setHyperlinkText("More")
            propertySyntaxLink!!.addHyperlinkListener(object : HyperlinkAdapter() {
                override fun hyperlinkActivated(e: HyperlinkEvent) {
                    try {
                        Desktop.getDesktop().browse(URI.create("https://edvin.gitbooks.io/tornadofx-guide/content/part2/Property%20Delegates.html#alternative-property-syntax"))
                    } catch (e1: IOException) {
                        e1.printStackTrace()
                    }
                }
            })
        }
    }
}